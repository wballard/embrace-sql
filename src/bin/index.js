#!/usr/bin/env node

// bootstrap typescript from the command line without a prebuild
require("ts-node").register({
  /* options */
});
// eslint-disable-next-line @typescript-eslint/no-var-requires
const path = require("path");

require(path.join(__dirname, "..", "cli", "index"));
