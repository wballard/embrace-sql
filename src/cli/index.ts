import { program } from "commander";
import pj from "../../package.json";
import init from "./init";
import start from "./start";
import migrate from "./migrate";
import { install } from "../structured-console";

/**
 * Our little command line interface. This can be used from Docker -- which is the
 * intended use, or with `yarn cli` for development and testing.
 */
const main = async (): Promise<void> => {
  install();
  program.version(pj.version);
  program.description(`Command line interface to EmbraceSQL.`);
  program.addCommand(init);
  program.addCommand(start);
  program.addCommand(migrate);
  // GO!
  await program.parseAsync(process.argv);
};
main();
