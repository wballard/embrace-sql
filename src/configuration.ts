import { cosmiconfig } from "cosmiconfig";
import Url from "url-parse";
import { generateFromTemplates } from "./generator";
import { LogLevel } from "./structured-console";
import path from "path";

/**
 * Named URLs to databases.
 */
type Databases = {
  [index: string]: Url;
};

/**
 * The all important root configuration. This tells EmbraceSQL which databases to manage.
 */
export type Configuration = {
  /**
   * The root directory used to start this config.
   */
  embraceSQLRoot: string;
  /**
   * URL identifying Kafka topic or REST endpoing to post commands.
   */
  twinCommandsTo?: string;
  /**
   * All available databases.
   */
  databases?: Databases;
  /**
   * Logging control
   */
  logLevels: LogLevel[];
  /**
   * Flag for embedded mode. This will be force set by the embedded cli.
   */
  embedded: boolean;
  /**
   * Optional name used in bootsrap code generation.
   */
  name?: string;
};

/**
 * Load up a configuration object.
 *
 * This will interpolate environment variables, build and object and hand it back.
 *
 * Will look in the current directory or a environment variable set root.
 */
export const loadConfiguration = async (
  root: string,
  embedded = false
): Promise<Configuration> => {
  // default name based on the path
  const name = root.split(path.sep).slice().pop() || "default";
  // run the root generation templates, gives you something to work
  // with even if you start in an empty directory so that system 'always works'
  await generateFromTemplates(
    {
      configuration: {
        name,
        embedded,
        embraceSQLRoot: root,
        logLevels: ["info", "error"],
      },
      databases: undefined,
      directQueryExecutors: {},
      close: (): Promise<void> => {
        return;
      },
    },
    "default"
  );
  // TODO env var substition loader hook
  // going with cosmic config -- even though this is just doing YAML for the moment
  const explorer = cosmiconfig("embracesql", {
    searchPlaces: ["embracesql.yaml", "embracesql.yml"],
  });
  const result = await explorer.search(root);
  const config = result.config as Configuration;
  config.embraceSQLRoot = root;
  // pop in some types so we are working with actual URLs
  const databases = Object.fromEntries(
    Object.keys(result.config.databases).map((databaseName) => [
      databaseName,
      new Url(result.config.databases[databaseName]),
    ])
  );
  return {
    ...config,
    logLevels: ["error"] || result.config.logLevels,
    databases,
  };
};
