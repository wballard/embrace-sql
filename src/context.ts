import { Configuration } from "./configuration";
import { embraceDatabases } from "./database-engines";
import { embraceEventHandlers } from "./event-handlers";
import {
  Database,
  SQLModule,
  SQLColumnMetadata,
  SQLParameters,
  SQLRow,
  SQLModuleDirectExecutors,
} from "./shared-context";
import { AST } from "node-sql-parser";
import { SQLModuleInternal } from "./event-handlers/sqlmodule-pipeline";

/**
 * Keep track of individual migration files with this type.
 */
export type MigrationFile = {
  name: string;
  content: string;
};

/**
 * This is the tree of paths derived from SQL files on disk. This is in
 * a compressed path format, so each key can have / in it.
 */
export type SQLModules = {
  [index: string]: SQLModule;
};

/**
 * A single instance of a database for use internally.
 */
export type DatabaseInternal = Database & {
  /**
   * All modules for this database.
   */
  SQLModules: SQLModules;
  /**
   * Execute the sql module query on this database, and
   * promise some result.
   *
   * @param SQLModule - execute this module, returning results
   * @param parameters - name value pairs are the passed parameters
   */
  execute: (
    sqlModule: SQLModule,
    parameters?: SQLParameters
  ) => Promise<SQLRow[]>;
  /**
   * Analyze the passed module and determine the resultset type(s).
   */
  analyze: (sqlModule: SQLModuleInternal) => Promise<SQLColumnMetadata[]>;
  /**
   * Parse out the SQL.
   */
  parse: (SQLModule) => AST[] | AST;
  /**
   * Do a migration.
   */
  migrate: (migrationFile: MigrationFile) => Promise<void>;
  /**
   * Clean close.
   */
  close: () => Promise<void>;
};

/**
 * All the databases from the internal point of view.
 */
export type AllDatabasesInternal = {
  [index: string]: DatabaseInternal;
};

/**
 * The root context is the context of configuration, databases, and sql modules and is
 * used to drive code generation and runtime execution 'in the engine' of EmbraceSQL
 *
 * The type here is a bit different from the context used in handlers, it has more metadata!
 */
export type InternalContext = SQLModuleDirectExecutors & {
  /**
   * The configuration used to build this context.
   */
  configuration: Configuration;
  /**
   * All configured databases, by name.
   */
  databases: AllDatabasesInternal;
  /**
   * Close all DB connections.
   */
  close: () => Promise<void>;
};

/**
 * With a configuration in hand, set up a new rootContext.
 *
 * This is built to be called -- repeatedly if needed. The idea is you can watch, and
 * rebuild a whole new context as needed -- swapping the root context at runtime to
 * hot-reconfigure the system without worrying about any state leaking.
 *
 * @param configuration - build a root context from this configuration.
 */
export const buildInternalContext = async (
  configuration: Configuration
): Promise<InternalContext> => {
  const databases = await embraceDatabases(configuration);
  const internalContext = {
    configuration,
    databases,
    directQueryExecutors: {},
    close: async (): Promise<void> => {
      const waitForThem = Object.values(databases).map((database) =>
        database.close()
      );
      await Promise.all(waitForThem);
    },
  };
  // need the database first, their connections are used
  // to mine metadata
  await embraceDatabases(configuration);
  await embraceEventHandlers(internalContext);
  return internalContext;
};
