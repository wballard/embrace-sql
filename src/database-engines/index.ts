import embraceSQLite from "./sqlite";
import { AllDatabasesInternal } from "../context";
import { DatabaseInternal } from "../context";
import pLimit from "p-limit";
import {
  SQLModule,
  SQLColumnMetadata,
  SQLRow,
  SQLParameters,
} from "../shared-context";
import { SQLModuleInternal } from "../event-handlers/sqlmodule-pipeline";
import { Configuration } from "../configuration";

/**
 * Serialize database use per process as we really only have one connection.
 */
const oneAtATime = pLimit(1);

/**
 * Embrace a database, bringing it into context.
 *
 * TODO: this needs a wrapper around each database to try to heal itself
 * and reconnect, or to fault and let the container worker process end.
 */
const embraceSingleDatabase = async (
  configuration: Configuration,
  databaseName: string
): Promise<DatabaseInternal> => {
  const dbUrl = configuration?.databases[databaseName];
  switch (dbUrl.protocol.split(":")[0].toLowerCase()) {
    case "sqlite":
      return embraceSQLite(configuration, databaseName);
    default:
      return undefined;
  }
};

/**
 * Every database in the configuraton is embraced and brought into the context.
 */
export const embraceDatabases = async (
  configuration: Configuration
): Promise<AllDatabasesInternal> => {
  // name value pairs inside the root context -- there isn't a strongly
  // typed generated context type available to the generator itself
  const databases = {};
  Object.keys(configuration.databases).forEach(async (databaseName) => {
    const database = await embraceSingleDatabase(configuration, databaseName);
    databases[databaseName] = {
      ...database,
      // here is wrapping the individual database driver execute and analyze
      // with a throttled promise limit -- since we have only one connection
      // these are not re-entrant
      execute: async (
        sqlModule: SQLModule,
        parameters?: SQLParameters
      ): Promise<SQLRow[]> => {
        const results = await oneAtATime(() =>
          database.execute(sqlModule, parameters)
        );
        return results as SQLRow[];
      },
      analyze: async (
        sqlModule: SQLModuleInternal
      ): Promise<SQLColumnMetadata[]> => {
        const results = await oneAtATime(() => database.analyze(sqlModule));
        return results as SQLColumnMetadata[];
      },
    };
  });
  return databases;
};
