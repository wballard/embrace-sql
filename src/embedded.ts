import { InternalContext, buildInternalContext } from "./context";
import { loadConfiguration } from "./configuration";

/**
 * Given a root directory, return an embedded context ready to be used with
 * and inprocess client.
 */
export const buildEmbeddedContext = async (
  root: string
): Promise<InternalContext> => {
  const configuration = await loadConfiguration(root);
  return buildInternalContext(configuration);
};
