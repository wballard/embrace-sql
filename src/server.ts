import { InternalContext } from "./context";
import Koa from "koa";
import bodyparser from "koa-bodyparser";
import OpenAPIBackend from "openapi-backend";
import YAML from "yaml";
import readFile from "read-file-utf8";
import path from "path";
import { HasContextualSQLModuleExecutors } from "./shared-context";

/**
 * Create a HTTP server exposing an OpenAPI style set of endpoints for each Database
 * and SQLModule.
 *
 * This does not actually start the server, it just hands you an instance that you
 * can start listening to later.
 *
 * @param rootContext - root server context with configuration
 * @param executionMap - context name to execution function mapping to actually 'run' a query
 */
export const createServer = async (
  rootContext: InternalContext & HasContextualSQLModuleExecutors
): Promise<Koa<Koa.DefaultState, Koa.DefaultContext>> => {
  const server = new Koa();

  // generated configuration is loaded up
  // TODO: the configuration is left on disk for current debugging -- but probably should be hidden
  const definition = YAML.parse(
    await readFile(
      path.join(rootContext.configuration.embraceSQLRoot, "openapi.yaml")
    )
  );

  const handlers = {};

  // go ahead and make a handler for both GET and POST
  // some of these GET handlers may not be connected at the OpenAPI layer
  // but a few extra functions isn't going to hurt anything
  Object.keys(rootContext.directQueryExecutors).forEach((contextName) => {
    handlers[`get__${contextName}`] = async (
      _openAPI,
      httpContext
    ): Promise<void> => {
      try {
        // parameters from the query
        const context = {
          parameters: httpContext.request.query,
          results: [],
        };
        await rootContext.contextualSQLModuleExecutors[contextName](context);
        httpContext.body = context.results;
        httpContext.status = 200;
      } catch (e) {
        console.error(e);
        httpContext.status = 500;
        httpContext.body = e;
      }
    };
    handlers[`post__${contextName}`] = async (
      _openAPI,
      httpContext
    ): Promise<void> => {
      try {
        // parameters from the body
        const context = {
          parameters: httpContext.request.body,
          results: [],
        };
        await rootContext.contextualSQLModuleExecutors[contextName](context);
        httpContext.body = context.results;
        httpContext.status = 200;
      } catch (e) {
        console.error(e);
        httpContext.status = 500;
        httpContext.body = e;
      }
    };
  });

  // merge up all the handlers just created with the OpenAPI definition
  // and we are ready to go -- this counts on OpenAPI doing some validation
  // before our handlers get called
  const api = new OpenAPIBackend({
    definition,
    handlers,
  });
  api.init();

  // use as koa middleware
  server.use(bodyparser());
  server.use((ctx) => api.handleRequest(ctx.request, ctx));

  // looks like we can get away without any code generation, and just building a
  // map of operation id based handlers
  // https://github.com/anttiviljami/openapi-backend/blob/master/examples/koa/index.js
  return server;
};
