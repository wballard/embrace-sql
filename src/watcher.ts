import { EventEmitter } from "events";
import { loadConfiguration } from "./configuration";
import { buildInternalContext, InternalContext } from "./context";
import chokidar from "chokidar";

/**
 * Need to be able to close this off to exit cleanly
 * in unit testing.
 */
export type CloseableEventEmitter = {
  emitter: EventEmitter;
  close: () => Promise<void>;
};
/**
 * Create a whole new context..
 */
export const reload = async (embraceSQLRoot): Promise<InternalContext> => {
  const configuration = await loadConfiguration(embraceSQLRoot);
  const newRootContext = await buildInternalContext(configuration);
  return newRootContext;
};

/**
 * Keep and eye on a root directory and rebuild it when files change.
 *
 * Note -- no need to worry about race conditions `buildRootContext` has
 * a throttle inside.
 */
export const watchRoot = (embraceSQLRoot: string): CloseableEventEmitter => {
  const emitter = new EventEmitter();

  // watch the whole directory
  const watcher = chokidar.watch([`${embraceSQLRoot}/**/*.sql`]);

  // this is super convenient to not drown in a pile of single file changes
  watcher.on("change", async () => {
    console.info("reload needed", embraceSQLRoot);
    const newRootContext = await reload(embraceSQLRoot);
    emitter.emit("reload", newRootContext);
  });
  return {
    emitter,
    close: (): Promise<void> => watcher.close(),
  };
};
