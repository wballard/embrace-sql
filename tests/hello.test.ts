/* eslint-disable @typescript-eslint/no-namespace */
import path from "path";
import fs from "fs-extra";
import { loadConfiguration, Configuration } from "../src/configuration";
import { buildInternalContext, InternalContext } from "../src/context";
import { createServer } from "../src/server";
import request from "supertest";
import readFile from "read-file-utf8";
import rmfr from "rmfr";
import { watchRoot } from "../src/watcher";
import http from "http";

declare global {
  namespace jest {
    interface Matchers<R> {
      toExist(): R;
    }
  }
}

/**
 * Hello world type tests, make sure the core configuration
 * and generation capabilities are working.
 *
 * This doesn't actually test that a server runs, or that you can
 * call APIs, -- just that configuration worked at all.
 */
describe("hello world configuration!", () => {
  let theConfig: Configuration = undefined;
  let rootContext: InternalContext;
  let root = "";
  let listening: http.Server;
  let callback;
  beforeAll(async () => {
    root = path.relative(process.cwd(), "./.tests/hello");
    // clean up
    await rmfr(root);
    // get the configuration and generate - let's do this just the once
    // and have a few tests that asser things happened
    const configuration = (theConfig = await loadConfiguration(root));
    rootContext = await buildInternalContext(configuration);
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const { decorateInternalContext } = require(path.join(
      process.cwd(),
      rootContext.configuration.embraceSQLRoot,
      "context"
    ));
    const server = await createServer(decorateInternalContext(rootContext));
    callback = server.callback();
    listening = server.listen(4567);
  });
  afterAll(async (done) => {
    listening.close(() => done());
  });
  expect.extend({
    toExist(fileName) {
      const fullPath = path.join(theConfig.embraceSQLRoot, fileName);
      const exists = fs.existsSync(fullPath);
      return exists
        ? { message: (): string => `${fullPath} exists`, pass: true }
        : { message: (): string => `${fullPath} does not exist`, pass: false };
    },
  });
  it("reads a config", async () => {
    expect(theConfig).toMatchSnapshot();
  });
  it("makes a default config for you", async () => {
    expect("embracesql.yaml").toExist();
  });
  it("makes a sqlite database for you", async () => {
    expect(theConfig.databases["default"].pathname).toExist();
  });
  it("makes a hello world sql for you", async () => {
    expect("default/hello.sql").toExist();
  });
  it("makes empty handlers for you", async () => {
    expect("default/before.ts").toExist();
    expect("default/hello.sql.before.ts").toExist();
    expect("default/hello.sql.after.ts").toExist();
    expect("default/after.ts").toExist();
    expect("default/hello.sql.afterError.ts").toExist();
  });
  it("exposes methods to run hello sql", async () => {
    expect(
      rootContext.databases["default"].SQLModules.hello.sql
    ).toMatchSnapshot();
  });
  it("generates an open api doc", async () => {
    expect("openapi.yaml").toExist();
    const content = await readFile(
      path.join(theConfig.embraceSQLRoot, "openapi.yaml")
    );
    expect(content).toMatchSnapshot();
  });
  it("generates a typed context object", async () => {
    expect("context.ts").toExist();
  });
  it("generates client library for you", async () => {
    expect("client/node/index.ts").toExist();
    expect("client/node-inprocess/index.ts").toExist();
    expect("client/browser/index.ts").toExist();
  });
  it("will run a query in context", async () => {
    const results = await rootContext.databases["default"].execute(
      rootContext.databases["default"].SQLModules["hello"]
    );
    expect(results).toMatchSnapshot();
  });
  it("will make a runnable server", async () => {
    const response = await request(callback).get("/default/hello");
    expect(response.text).toMatchSnapshot();
    // client
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const { EmbraceSQL } = require(path.join(
      process.cwd(),
      rootContext.configuration.embraceSQLRoot,
      "client",
      "node"
    ));
    const client = EmbraceSQL("http://localhost:4567");
    expect(await client.databases.default.hello.sql()).toMatchSnapshot();
  });
  it("will make an embeddable engine", async () => {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const { EmbraceSQL } = require(path.join(
      process.cwd(),
      rootContext.configuration.embraceSQLRoot,
      "client",
      "node-inprocess"
    ));
    const client = EmbraceSQL(rootContext);
    expect(await client.databases.default.hello.sql()).toMatchSnapshot();
  });
  it("will watch for changes and create a new context", async (done) => {
    const watcher = watchRoot(root);
    watcher.emitter.on("reload", async (newContext: InternalContext) => {
      // it's a new object
      expect(newContext).not.toBe(rootContext);
      // and it has the values we expect
      expect(newContext).toMatchSnapshot();
      await watcher.close();
      // and let Jest finish
      done();
    });
    // adding a new file should trigger the watcher
    // calling back to the event, which should tell jest we are all done
    await fs.outputFile(
      path.join(theConfig.embraceSQLRoot, "default", "yo.sql"),
      "SELECT 'yo'"
    );
  });
  it("will make migrations directories", async () => {
    expect("migrations/default").toExist();
  });
});
