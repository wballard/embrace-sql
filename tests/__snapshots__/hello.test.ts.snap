// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hello world configuration! exposes methods to run hello sql 1`] = `"SELECT \\"hello world\\""`;

exports[`hello world configuration! generates an open api doc 1`] = `
"# THIS FILE IS GENERATED -- edits will be lost

openapi: 3.0.0
info:
  title: --
  description: --
  version: --

paths:
  /default/hello:
    get:
      operationId: get__default_hello
      description: Read data
      responses:
        \\"200\\":
          description: Sucess
          content:
            application/json:
              schema:
                type: array
                items:
                  type: object
                  properties:
                    hello_world:
                      type: string
    post:
      operationId: post__default_hello
      description: Read data
      requestBody:
        required: true
        content:
          application/json:
            schema:
              type: object
      responses:
        \\"200\\":
          description: Sucess
          content:
            application/json:
              schema:
                type: array
                items:
                  type: object
                  properties:
                    hello_world:
                      type: string
"
`;

exports[`hello world configuration! reads a config 1`] = `
Object {
  "databases": Object {
    "default": Object {
      "auth": "",
      "hash": "",
      "host": "",
      "hostname": "",
      "href": "sqlite:embracesql.db",
      "origin": "null",
      "password": "",
      "pathname": "embracesql.db",
      "port": "",
      "protocol": "sqlite:",
      "query": "",
      "slashes": false,
      "username": "",
    },
  },
  "embedded": false,
  "embraceSQLRoot": ".tests/hello",
  "logLevels": Array [
    "error",
  ],
}
`;

exports[`hello world configuration! will make a runnable server 1`] = `"[{\\"\\\\\\"hello world\\\\\\"\\":\\"hello world\\"}]"`;

exports[`hello world configuration! will make a runnable server 2`] = `
Array [
  Object {
    "\\"hello world\\"": "hello world",
  },
]
`;

exports[`hello world configuration! will make an embeddable engine 1`] = `
Array [
  Object {
    "\\"hello world\\"": "hello world",
  },
]
`;

exports[`hello world configuration! will run a query in context 1`] = `
Array [
  Object {
    "\\"hello world\\"": "hello world",
  },
]
`;

exports[`hello world configuration! will watch for changes and create a new context 1`] = `
Object {
  "close": [Function],
  "configuration": Object {
    "databases": Object {
      "default": Object {
        "auth": "",
        "hash": "",
        "host": "",
        "hostname": "",
        "href": "sqlite:embracesql.db",
        "origin": "null",
        "password": "",
        "pathname": "embracesql.db",
        "port": "",
        "protocol": "sqlite:",
        "query": "",
        "slashes": false,
        "username": "",
      },
    },
    "embedded": false,
    "embraceSQLRoot": ".tests/hello",
    "logLevels": Array [
      "error",
    ],
  },
  "databases": Object {
    "default": Object {
      "SQLModules": Object {
        "hello": Object {
          "afterHandlerPaths": Array [
            "default",
          ],
          "ast": Object {
            "columns": Array [
              Object {
                "as": null,
                "expr": Object {
                  "type": "string",
                  "value": "hello world",
                },
              },
            ],
            "distinct": null,
            "from": null,
            "groupby": null,
            "having": null,
            "limit": null,
            "options": null,
            "orderby": null,
            "type": "select",
            "where": null,
            "with": null,
          },
          "beforeHandlerPaths": Array [
            "default",
          ],
          "cacheKey": "4726f48a1467bab00568c7c5b8610040",
          "canModifyData": false,
          "contextName": "default_hello",
          "fullPath": ".tests/hello/default/hello.sql",
          "namedParameters": Array [],
          "restPath": "default/hello",
          "resultsetMetadata": Array [
            Object {
              "name": "hello_world",
              "type": "string",
            },
          ],
          "sql": "SELECT \\"hello world\\"",
        },
        "yo": Object {
          "afterHandlerPaths": Array [
            "default",
          ],
          "ast": Object {
            "columns": Array [
              Object {
                "as": null,
                "expr": Object {
                  "type": "single_quote_string",
                  "value": "yo",
                },
              },
            ],
            "distinct": null,
            "from": null,
            "groupby": null,
            "having": null,
            "limit": null,
            "options": null,
            "orderby": null,
            "type": "select",
            "where": null,
            "with": null,
          },
          "beforeHandlerPaths": Array [
            "default",
          ],
          "cacheKey": "54d6f4897bf10f8219c6fc94e11bf7a4",
          "canModifyData": false,
          "contextName": "default_yo",
          "fullPath": ".tests/hello/default/yo.sql",
          "namedParameters": Array [],
          "restPath": "default/yo",
          "resultsetMetadata": Array [
            Object {
              "name": "yo",
              "type": "string",
            },
          ],
          "sql": "SELECT 'yo'",
        },
      },
      "analyze": [Function],
      "close": [Function],
      "execute": [Function],
      "migrate": [Function],
      "name": "default",
      "parse": [Function],
      "transactions": Object {
        "begin": [Function],
        "commit": [Function],
        "rollback": [Function],
      },
    },
  },
  "directQueryExecutors": Object {
    "default_hello": [Function],
    "default_yo": [Function],
  },
}
`;
